server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,442 @@
+                    logger.info("successfully cleared voting config after decommissioning");
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    logger.debug(new ParameterizedMessage("failure in clearing voting config exclusion after decommissioning"), exp);
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
+                }
+
+                @Override
+                public ClearVotingConfigExclusionsResponse read(StreamInput in) throws IOException {
+                    return new ClearVotingConfigExclusionsResponse(in);
+                }
+            }
+        );
+    }
+
+    /**
+     * Registers new decommissioned attribute metadata in the cluster state with {@link DecommissionStatus#DECOMMISSION_INIT}
+     * <p>
+     * This method can be only called on the cluster-manager node. It tries to create a new decommissioned attribute on the master
+     * and if it was successful it adds new decommissioned attribute to cluster metadata.
+     * <p>
+     * This method ensures that request is performed only on eligible cluster manager node
+     *
+     * @param decommissionAttribute register decommission attribute in the metadata request
+     * @param listener              register decommission listener
+     */
+    private void registerDecommissionAttribute(
+        final DecommissionAttribute decommissionAttribute,
+        final ActionListener<ClusterStateUpdateResponse> listener
+    ) {
+        if (!transportService.getLocalNode().isClusterManagerNode()
+            || nodeHasDecommissionedAttribute(transportService.getLocalNode(), decommissionAttribute)) {
+            throw new NotClusterManagerException(
+                "node ["
+                    + transportService.getLocalNode().toString()
+                    + "] not eligible to execute decommission request. Will retry until timeout."
+            );
+        }
+        clusterService.submitStateUpdateTask(
+            "put_decommission [" + decommissionAttribute + "]",
+            new ClusterStateUpdateTask(Priority.URGENT) {
+                @Override
+                public ClusterState execute(ClusterState currentState) {
+                    logger.info(
+                        "registering decommission metadata for attribute [{}] with status as [{}]",
+                        decommissionAttribute.toString(),
+                        DecommissionStatus.DECOMMISSION_INIT
+                    );
+                    Metadata metadata = currentState.metadata();
+                    Metadata.Builder mdBuilder = Metadata.builder(metadata);
+                    DecommissionAttributeMetadata decommissionAttributeMetadata = metadata.custom(DecommissionAttributeMetadata.TYPE);
+                    ensureNoAwarenessAttributeDecommissioned(decommissionAttributeMetadata, decommissionAttribute);
```

Comment:
should we do this validations beforehand ? It would be ideal to this before even sending transport request .